File: DTCore5.0-exp/DTInterop/DataTools.Interop.Bluetooth/BluetoothDeviceInfo.vb
'' ************************************************* ''
'' DataTools Visual Basic Utility Library - Interop
''
'' Module: BlueTooth information (TO DO)
'' 
'' Copyright (C) 2011-2020 Nathan Moschkin
'' All Rights Reserved
''
'' Licensed Under the Microsoft Public License   
'' ************************************************* ''


Imports CoreCT.Memory
Imports DataTools.Interop.Native
Imports System.ComponentModel
Imports System.Reflection
Imports System.Collections.ObjectModel
Imports System.Windows.Media.Imaging
Imports DataTools.Interop.Printers
Imports DataTools.SystemInformation
Imports DataTools.Interop.Desktop
Imports System.Runtime.InteropServices


'' TO DO!  
#Region "Bluetooth Device Info"

Public Class BluetoothDeviceInfo
    Inherits DeviceInfo

    Protected _bthaddr As BLUETOOTH_ADDRESS

    Protected _service As String
    Protected _major As String
    Protected _minor As String

    Protected _btClass As UInteger

    Protected _isradio As Boolean

    Public Overridable Property IsRadio As Boolean
        Get
            Return _isradio
        End Get
        Friend Set(value As Boolean)
            _isradio = value
        End Set
    End Property

    Public Overridable Property BluetoothDeviceClass As UInteger
        Get
            Return _btClass
        End Get
        Friend Set(value As UInteger)
            _btClass = value

            Dim svc As UShort,
                maj As UShort,
                min As UShort

            ParseClass(_btClass, svc, maj, min)

            _service = PrintServiceClass(svc)
            _major = PrintMajorClass(maj)
            _minor = PrintMinorClass(maj, min)

        End Set
    End Property


    ''' <summary>
    ''' Return the Bluetooth MAC address for this radio
    ''' </summary>
    ''' <returns></returns>
    Public Overridable Property BluetoothAddress As BLUETOOTH_ADDRESS
        Get
            Return _bthaddr
        End Get
        Friend Set(value As BLUETOOTH_ADDRESS)
            _bthaddr = value
        End Set
    End Property


    Public Overridable Property BluetoothServiceClasses As String
        Get
            Return _service
        End Get
        Protected Set(value As String)
            _service = value
        End Set
    End Property


    Public Overridable Property BluetoothMajorClasses As String
        Get
            Return _major
        End Get
        Protected Set(value As String)
            _major = value
        End Set
    End Property


    Public Overridable Property BluetoothMinorClasses As String
        Get
            Return _minor
        End Get
        Protected Set(value As String)
            _minor = value
        End Set
    End Property

    Public Overrides ReadOnly Property UIDescription As String
        Get
            If String.IsNullOrEmpty(FriendlyName) Then Return MyBase.UIDescription Else Return FriendlyName
        End Get
    End Property


    Public Shared Sub ShowBluetoothSettings()

        Dim shex As New SHELLEXECUTEINFO
        shex.cbSize = Marshal.SizeOf(shex)
        shex.nShow = SW_SHOW
        shex.hInstApp = Process.GetCurrentProcess.Handle
        'shex.hWnd = 
        'shex.lpVerb = "properties"

        '' Set the parsing name exactly this way.
        shex.lpFile = "ms-settings:bluetooth"

        shex.fMask = SEE_MASK_ASYNCOK Or SEE_MASK_FLAG_DDEWAIT Or SEE_MASK_UNICODE
        ShellExecuteEx(shex)

        'Shell("ms-settings:bluetooth")
    End Sub


    Public Overrides Function ToString() As String
        If String.IsNullOrEmpty(FriendlyName) Then Return MyBase.ToString Else Return FriendlyName
    End Function

End Class

#End Region

File: DataTools5/DataTools.Hardware/Bluetooth/BluetoothDeviceInfo.cs
// ************************************************* ''
// DataTools C# Native Utility Library For Windows - Interop
//
// Module: BlueTooth information (TO DO)
// 
// Copyright (C) 2011-2020 Nathan Moschkin
// All Rights Reserved
//
// Licensed Under the Microsoft Public License   
// ************************************************* ''


using System.Diagnostics;
using System.Runtime.InteropServices;
using DataTools.Win32Api;

namespace DataTools.Hardware
{


    // TO DO!  
    /* TODO ERROR: Skipped RegionDirectiveTrivia */
    public class BluetoothDeviceInfo : DeviceInfo
    {
        protected BLUETOOTH_ADDRESS _bthaddr;
        protected string _service;
        protected string _major;
        protected string _minor;
        protected uint _btClass;
        protected bool _isradio;

        public virtual bool IsRadio
        {
            get
            {
                return _isradio;
            }

            internal set
            {
                _isradio = value;
            }
        }

        public virtual uint BluetoothDeviceClass
        {
            get
            {
                return _btClass;
            }

            internal set
            {
                _btClass = value;
                var svc = default(ushort);
                var maj = default(ushort);
                var min = default(ushort);
                Bluetooth.ParseClass(_btClass, ref svc, ref maj, ref min);
                _service = Bluetooth.PrintServiceClass(svc);
                _major = Bluetooth.PrintMajorClass(maj);
                _minor = Bluetooth.PrintMinorClass(maj, min);
            }
        }


        /// <summary>
    /// Return the Bluetooth MAC address for this radio
    /// </summary>
    /// <returns></returns>
        public virtual BLUETOOTH_ADDRESS BluetoothAddress
        {
            get
            {
                return _bthaddr;
            }

            internal set
            {
                _bthaddr = value;
            }
        }

        public virtual string BluetoothServiceClasses
        {
            get
            {
                return _service;
            }

            protected set
            {
                _service = value;
            }
        }

        public virtual string BluetoothMajorClasses
        {
            get
            {
                return _major;
            }

            protected set
            {
                _major = value;
            }
        }

        public virtual string BluetoothMinorClasses
        {
            get
            {
                return _minor;
            }

            protected set
            {
                _minor = value;
            }
        }

        public override string UIDescription
        {
            get
            {
                if (string.IsNullOrEmpty(FriendlyName))
                    return base.UIDescription;
                else
                    return FriendlyName;
            }
        }

        public static void ShowBluetoothSettings()
        {
            var shex = new SHELLEXECUTEINFO();
            shex.cbSize = Marshal.SizeOf(shex);
            shex.nShow = User32.SW_SHOW;
            shex.hInstApp = Process.GetCurrentProcess().Handle;
            // shex.hWnd = 
            // shex.lpVerb = "properties"

            // Set the parsing name exactly this way.
            shex.lpFile = "ms-settings:bluetooth";
            shex.fMask = User32.SEE_MASK_ASYNCOK | User32.SEE_MASK_FLAG_DDEWAIT | User32.SEE_MASK_UNICODE;
            User32.ShellExecuteEx(ref shex);

            // Shell("ms-settings:bluetooth")
        }

        public override string ToString()
        {
            if (string.IsNullOrEmpty(FriendlyName))
                return base.ToString();
            else
                return FriendlyName;
        }
    }
}

/* TODO ERROR: Skipped EndRegionDirectiveTrivia */